EIPS/eip-draft.md

```diff
@@ -0,0 +1,226 @@
+
+3. Leverages existing infrastructure: 
+Builds on EIP-2930 access lists and EIP-2929 access costs, requiring minimal protocol changes.
+
+4. Explicit opt-in: 
+Transactions must explicitly indicate which addresses to check. This prevents unexpected behavior and gives users/wallets control over gas optimization.
+
+5. Forward compatibility: 
+Pre-fork nodes ignore `checkCodeHash` and never grant discounts. Post-fork, all nodes enforce identical behavior. Wallets can optionally add the field to optimize gas, but its absence doesn't invalidate transactions.
+
+6. Avoiding have a code-root for state:
+At this point, clients handle code storage on their own ways. They don't have any consensus on the deployed existing codes (besides that all of the ones referenced in account's codehash fields exist).
+Changing this seems a lot more complex and unnecessary.
+
+### Same-Block Deployments
+
+Sequential transaction execution ensures that a deployment storing new code makes it visible to later transactions in the same block:
+
+1. Transaction `T_A` deploys bytecode `B` at address `X`
+   - Pays full `GAS_CODE_DEPOSIT * L` (no prior contract has this bytecode)
+   - Code is stored under hash `H = keccak256(B)`
+
+2. Later transaction `T_B` in the same block deploys the same bytecode `B`:
+   - `T_B` includes address `X` in its access list with `checkCodeHash: true`
+   - When `T_B` executes, `W` is built from the current state (including `T_A`'s changes)
+   - Since `X` now exists, `W` contains `H`
+   - `T_B`'s deployment gets the discount
+
+> While this only tries to formalize the behaviour, it's important to remark that this kind of behaviour is complex. As it requires control over tx ordering in order to abuse. And Builders can't modify the Acess List as it is already signed with the Tx. Nevertheless, this could happen, thus is formalized here. 
+
+### Edge Case: Simultaneous New Deployments
+
+If two transactions in the same block both deploy identical new bytecode and neither references an existing contract with that bytecode in their access lists, both will pay full `GAS_CODE_DEPOSIT * L`. This is acceptable because:
+
+- The first deployment cannot be known at transaction construction time
+- This scenario is extremely rare in practice
+- The complexity of special handling is not worth the minimal benefit
+
+## Backwards Compatibility
+
+This proposal requires a scheduled network upgrade but is designed to be forward-compatible with existing transactions.
+
+**Transaction compatibility:**
+- Transactions with `checkCodeHash` fields are syntactically valid both pre-fork and post-fork
+- Pre-fork: The field is ignored; all deployments pay full costs
+- Post-fork: The field determines deduplication behavior
+
+**Wallet and tooling updates:**
+- RPC methods like `eth_estimateGas` MUST account for potential deduplication discounts
```

Comment:
is it as must though? worst case is, they declare a higher gas limit but the extra is refunded.